### PR TITLE
*: Refactor controller implementation

### PR DESCRIPTION
**Pull request by timflannagan**
Signed-off-by: timflannagan <timflannagan@gmail.com>

### CODE REVIEW

File: controllers/platformoperator_controller.go [hunk 1]
```diff
@@ -100,162 +59,32 @@ func (r *PlatformOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
-			cb = append(cb, b)
-		}
-	}
-	if len(errors) != 0 {
-		agg := utilerror.NewAggregate(errors)
-		log.Error(agg, "failed to build catalog snapshot")
-		return ctrl.Result{}, agg
+	desiredBundle, err := r.Sourcer.Source(ctx, po)
+	if err != nil {
+		log.Error(err, "failed to source content for platform operator")
```
Comment (exdx) [inline controllers/platformoperator_controller.go:69]: nit: should we include the name of the PO in a formatted string in the error log for more visibility? 
```
log.Error(err, "failed to source content for platform operator", po.GetName())
```

File: controllers/platformoperator_controller.go [hunk 2]
```diff
@@ -100,162 +59,32 @@ func (r *PlatformOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
-	if len(cb) == 0 {
-		log.Info("failed to find any candidate bundles")
-		return ctrl.Result{}, nil
-	}
-
-	latestBundle, err := cb.latest()
-	if err != nil || latestBundle == nil {
-		log.Info("failed to find the bundle with the highest semver")
+	if err := r.Applier.Apply(ctx, po, desiredBundle); err != nil {
+		log.Error(err, "failed to apply the desired bundle")
```
Comment (exdx) [inline controllers/platformoperator_controller.go:73]: nit: same here, maybe more context for the user?
```
log.Error(err, "failed to apply the desired bundle", desiredBundle.GetName(), "for platform operator", po.GetName())

File: internal/sourcer/source.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+type Bundle struct {
+	Version  string
+	Image    string
+	Replaces string
+	Skips    []string
+}
+
+func (b Bundle) String() string {
+	return fmt.Sprintf("Version: %s; Image: %s; Repliaces; %s", b.Version, b.Image, b.Replaces)
```
Comment (exdx) [inline internal/sourcer/source.go:27]:
```suggestion
	return fmt.Sprintf("Version: %s; Image: %s; Replaces; %s", b.Version, b.Image, b.Replaces)
```

**Review by exdx — APPROVED** [review]
LGTM. Having these interfaces here so early on will make iterating a lot nicer.

File: controllers/platformoperator_controller.go (hunk 1 above)
Comment (timflannagan) [reply controllers/platformoperator_controller.go:69]: We should already have this information from the logger given we create a logger from the Reconcile ctx provided to us: `log := logr.FromContext(ctx)`.

File: controllers/platformoperator_controller.go (hunk 2 above)
Comment (timflannagan) [reply controllers/platformoperator_controller.go:73]: We should already have this information from the logger given we create a logger from the Reconcile ctx provided to us: `log := logr.FromContext(ctx)`.

File: internal/sourcer/source.go (hunk 3 above)
Comment (timflannagan) [reply internal/sourcer/source.go]: Whoops - this was a debug artifact, but I'll keep it in anyways. Updated.

**Pull request merged**